module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,167 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+import org.springframework.lang.Nullable;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@NoArgsConstructor
+@Repository
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class GiftCertificateDaoImpl implements GiftCertificateDao {
+
+    @Autowired JdbcTemplate jdbcTemplate;
+    @Autowired NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    @Autowired @Qualifier("tag_giftCertificate") SimpleJdbcInsert simpleJdbcInsert;
+
+    @Override
+    @Transactional(propagation = Propagation.REQUIRED)
+    public void save(GiftCertificate certificate, List<Tag> tags) {
+        saveGiftCertificate(certificate);
+        saveReferencesBetweenCertificatesAndTags(get(certificate.getName()), tags);
+    }
+
+    private void saveGiftCertificate(GiftCertificate certificate) {
+        namedParameterJdbcTemplate.update(
+                "INSERT INTO giftCertificate (name, description, price, duration) " +
```

Comment:
Please consider using single naming convention for database tables and columns. Either camelCase or snake_case, not both. snake_case seems to be conventional these days.